File: python3-bottle/infix.py
from collections import deque

import rpn

PRECEDENCE = {
    '^': 3,
    '*': 2,
    '/': 2,
    '-': 1,
    '+': 1,
    '(': 0
}

class InfixException(Exception):
    pass

def convert(tokens):
    stack = deque()

    for token in tokens:
        if token.isdigit():
            yield token
        elif token == '(':
            stack.append(token)
        elif token == ')':
            while True:
                if len(stack) == 0:
                    raise InfixException('Stack underflow')
                operator = stack.pop()
                if operator == '(':
                    break
                yield operator
        else:
            token_precedence = PRECEDENCE.get(token)
            while True:
                if len(stack) > 0:
                    stack_precedence = PRECEDENCE.get(stack[-1])
                else:
                    stack_precedence = 0
                if token_precedence > stack_precedence:
                    stack.append(token)
                    break
                elif token_precedence < stack_precedence:
                    yield stack.pop()
                else:
                    yield stack.pop()
                    stack.append(token)
                    break

    while len(stack) > 0:
        yield stack.pop()

def evaluate(tokens):
    return rpn.evaluate(convert(tokens))

File: node-express/infix.js
'use strict';

const _ = require('underscore');
const rpn = require('./rpn');

const precedence = {
    '^': 3,
    '*': 2,
    '/': 2,
    '-': 1,
    '+': 1,
    '(': 0
};

let convertInfix = function(tokens) {
    let stack = [];
    let output = [];

    tokens.forEach(function(token) {
        if (_.isNumber(token)) {
            output.push(token);
        } else if (token === '(') {
            stack.push(token);
        } else if (token === ')') {
            while (true) {
                if (stack.length <= 0) {
                    throw 'Stack underflow';
                }
                let operator = stack.pop();
                if (operator === '(') {
                    break;
                }
                output.push(operator);
            }
        } else {
            let tokenPrecedence = precedence[token];
            while (true) {
                let stackPrecedence = precedence[stack[stack.length-1]] || 0;
                if (tokenPrecedence > stackPrecedence) {
                    stack.push(token);
                    break;
                } else if (tokenPrecedence < stackPrecedence) {
                    output.push(stack.pop());
                } else {
                    output.push(stack.pop());
                    stack.push(token);
                    break;
                }
            }
        }
    });

    while (stack.length > 0) {
        output.push(stack.pop());
    }

    return output;
};

let evaluateInfix = _.compose(rpn.evaluateRpn, convertInfix);

module.exports = {
    convertInfix,
    evaluateInfix
};
